File: src/main/java/com/kevin/student/Student.java
package com.kevin.student;

public class Student {
    String id;
    public String name;
    public int english;
    public int math;
    static int pass = 60;

    public Student(String name, int english, int math) {
        this.name = name;
        this.english = english;
        this.math = math;
    }

    public void print() {
        int average = getAverage();

        System.out.println(name + "\t" + english +
                "\t" + math +
                "\t" + getAverage() + "\t" +
                ((average >= pass) ? "PASS" : "FAILED") + "\t" +
                ( switch (average / 10) {
                    case 10, 9 -> "A";
                    case 8, 7 -> "B";
                    case 6, 5 -> "C";
                    default -> "F";
                })
                );
//        char grading = 'F';
//        switch (average / 10) {
//            case 10:
//                case 9:
//                    grading = 'A';
//                    break;
//            case 8:
//                case 7:
//                    grading = 'B';
//                    break;
//            case 6:
//                case 5:
//                    grading = 'C';
//                    break;
//            default:
//                grading = 'F';
//        }

//        if (average >= 60) {
//            System.out.println("\tPASS");
//        } else {
//            System.out.println("\tFAILED");
//        }
    }

    public int getAverage() {
        return (int) Math.floor((double)(english + math) / 2);
    }

    public int highest() {
        return Math.max(english, math);
    }
}

File: src/main/kotlin/com/kevin/student/Student.kt
package com.kevin.kotlin

import java.util.*

fun main() {
//    userInput()
    Student.pass = 55
    val stu = Student("Kevin", 44, 66);
    val stu2 = Student("Hank", 100, 100);
    val stu3 = Student("Jane", 30, 50);
    stu.print();
    stu2.print();
    stu3.print();
    val gstu = GraduateStudent("Jack", 55, 65, 60);
    gstu.print();
//    var test = 123
//    println("Test is ${test}")
    println("High score: ${stu.highest()}")
}

class GraduateStudent(name: String?, english: Int, math: Int, private var thesis: Int) : Student(name, english, math) {
    companion object {
        var pass = 70
    }
    override fun print() {
        println(
            "$name\t$english\t$math\t$thesis\t${getAverage()}\t${passOrFailed()}\t${grading()}"
        )
    }

    override fun passOrFailed() = if (getAverage() >= pass) "PASS" else "FAILED"
}

open class Student(var name: String?, var english: Int, var math: Int) {
    companion object {
        @JvmStatic  // 這個 annotation 可以讓 Java 程式碼呼叫 Kotlin 的 static method
        var pass = 60
        fun test() {
            println("testing")
        }
    }
    open fun print() {
//        print(
//            name + "\t" + english + "\t" + math +
//                    "\t" + getAverage() + "\t" +
//                    if (getAverage() >= 60) "PASS" else "FAILED"
//        )
//        println("\t" + grading())
        println(
            "$name\t$english\t$math\t${getAverage()}\t${passOrFailed()}\t${grading()}"
        )
    }

    open fun passOrFailed() = if (getAverage() >= pass) "PASS" else "FAILED"

    internal fun grading(): Char = when (getAverage()) {
        in 90..100 -> 'A'
        in 80..89 -> 'B'
        in 70..79 -> 'C'
        in 60..69 -> 'D'
        else -> 'F'
    }

    fun getAverage(): Int = (english + math) / 2

    fun nameCheck() = println(name?.length)

    fun highest(): Int =
        if (english > math) {
            println("english")
            english
        } else {
            println("math")
            math
        }

}

private fun userInput() {
    val scanner = Scanner(System.`in`)
    print("Please enter student's name: ")
//    var name = scanner.next()
    var name = null
    print("Please enter student's english: ")
    var english = scanner.nextInt()
    print("Please enter student's math: ")
    var math = scanner.nextInt()
    val stu = Student(name, english, math);
    stu.print();
    stu.nameCheck()
}


